### PR TITLE
feat: unambiguous model file command line options

### PR DESCRIPTION
**Pull request by jackgerrits**
- Rev version to 9.1.0 to ensure model file change is picked up
- Enable unambiguous option tokenization for model command lines
- Remove usage of Boost Program options for model command line processing 
- Enable string escaping and quoting when serializing and proper quote/escape command line splitting when reading
    - Any value that included a space in the past would actually break command line parsing from model files
- Loading a model file with the legacy style is still parsed based on the ambiguous rules of that format
- New models written in the new format

By changing the format of model command lines from:
```
--bool_switch --scalar_option value --scalar_option value with space
```

To:
```
--bool_switch --scalar_option=value --scalar_option="value with space"
```

It allows for clear association of values with option names. 

Requires: #3725

### CODE REVIEW

File: vowpalwabbit/config/cli_options_serializer.cc [hunk 1]
```diff
@@ -4,35 +4,52 @@
...
+cli_options_serializer::cli_options_serializer() : cli_options_serializer(false) {}
+cli_options_serializer::cli_options_serializer(bool escape) : m_escape(escape) {}
 
 template <typename T>
-void serialize(std::stringstream& output, const typed_option<T>& typed_option)
+void serialize(std::stringstream& output, const typed_option<T>& typed_option, bool escape)
 {
-  output << " --" << typed_option.m_name << " " << typed_option.value();
+  std::stringstream ss;
+  ss.precision(15);
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/config/cli_options_serializer.cc:20]: This is a carry over from when we used doubles in keep options for save resume state. A double has 15 decimal digits of precision

File: vowpalwabbit/parse_args.cc [hunk 2]
```diff
@@ -1438,93 +1439,88 @@ bool check_interaction_settings_collision(options_i& options, const std::string&
...
 
-void merge_options_from_header_strings(const std::vector<std::string>& strings, bool skip_interactions,
-    VW::config::options_i& options, bool& is_ccb_input_model)
-{
-  po::options_description desc("");
+bool is_long_option_like(VW::string_view token) { return token.find("--") == 0 && token.size() > 2; }
 
-  // Get list of options in file options std::string
-  po::parsed_options pos = po::command_line_parser(strings).options(desc).allow_unregistered().run();
+// The model file contains a command line but it has much greater constraints than the user supplied command line. These
```
Comment (jackgerrits) [inline vowpalwabbit/parse_args.cc:1444]: TODO: Update comment to reflect greater constraints of equals style syntax

File: vowpalwabbit/config/cli_options_serializer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/config/cli_options_serializer.cc:20]: Yeah will do. Wonder if we should reduce it now that we only use floats.

File: vowpalwabbit/parse_primitives.cc [hunk 3]
```diff
@@ -151,4 +155,45 @@ std::vector<std::string> split_command_line(const std::string& cmd_line)
...
+        ss << c;
+        break;
+    }
+  }
+  return ss.str();
+}
+
+bool contains_escapeable_chars(VW::string_view str)
+{
+  const std::set<char> escape_chars = {'\n', '\t', '\\', '"', '\''};
```
Comment (olgavrou) [inline vowpalwabbit/parse_primitives.cc:191]: `\r`?

File: vowpalwabbit/config/cli_options_serializer.h [hunk 4]
```diff
@@ -17,6 +17,7 @@ namespace config
 struct cli_options_serializer : options_serializer_i, typed_option_visitor
 {
   cli_options_serializer();
+  cli_options_serializer(bool escape);
```
Comment (olgavrou) [inline vowpalwabbit/config/cli_options_serializer.h:20]:
```suggestion
  explicit cli_options_serializer(bool escape);
```
?

File: vowpalwabbit/config/cli_options_serializer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (olgavrou) [reply vowpalwabbit/config/cli_options_serializer.cc:20]: would have to gate that behind the model version for compatibility right? Might be worth it, might be ok as is

File: vowpalwabbit/parse_regressor.cc [hunk 5]
```diff
@@ -387,7 +441,8 @@ void save_load_header(VW::workspace& all, io_buf& model_file, bool read, bool te
     }
     else
     {
-      VW::config::cli_options_serializer serializer;
+      // Escape values in output.
+      VW::config::cli_options_serializer serializer(true);
```
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/parse_regressor.cc:445]: Yeah I think that makes sense. Supplying false for this arg is essentially saying "I understand I may produce a command line that cannot be read back in"

File: vowpalwabbit/config/cli_options_serializer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jackgerrits) [reply vowpalwabbit/config/cli_options_serializer.cc:20]: Doesn't need to be gated. This is only affects newly produced models, for which every one of those uses floats and not doubles.

File: vowpalwabbit/parse_primitives.cc (hunk 3 above)
Comment (olgavrou) [reply vowpalwabbit/parse_primitives.cc:191]: Yeah I see what you are saying. Having a `\n` or a `\t` in a value of a cli arg is very particular

Comment (jackgerrits) [reply vowpalwabbit/parse_primitives.cc:191]: Going to reduce the scope of escaping to just \ ' "

**Comment by jackgerrits** [comment]
I'm going to split this into two changes:
- Refactor out Boost PO
- Change format to be be unambiguous

**Pull request closed without merging**